### PR TITLE
Social media website - Milestone: friendships v1

### PR DESCRIPTION
**Pull request by mahmoud717**
### In this milestone I did the following:

- created a model with associations and all requested features for friendships
- added links and routes for every action

### CODE REVIEW

File: app/views/users/_user.html.erb [hunk 1]
```diff
@@ -1,6 +1,21 @@
...
+        <%= link_to "Unfriend", unfriend_path(current_user.id, user.id), method: "POST" %>
+        <% elsif current_user.friend_requests.include?(user)%>
+        <%= link_to "confirm", confirmfriendrequest_path(Friendship.find_by(user_id: user.id).id) , method: "POST" %>
+        <%= link_to "cancel", cancelfriendrequest_path(Friendship.find_by(user_id: user.id).id) , method: "POST" %>
+        <% elsif user.friend_requests.include?(current_user)%>
+        <p> request sent
+        </p>
+        <% else %>
+        <%= link_to "Add Friend " , sendfriendrequest_path(current_user.id, user.id), method: "POST"%>
+        <% end %>
```
Comment (maelfosso) [inline app/views/users/_user.html.erb]: No logic in views, please.
Implement them into view helpers

**Review by maelfosso — CHANGES_REQUESTED** [review]
# Changes required 🔴 

Hello,

Thank you for the well-done job so far,
However, there are some missing requirements

1- The `rubocop.yml` file is different from the required one. Kindly follow this [link](https://github.com/microverseinc/linters-config/tree/master/ror) to set up the good one

2- There are no tests. 
It's required, as you can read it [here](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/tests.png), you must write **unit** and **integrations** tests using RSpec
Kindly add them, please.

For the unit tests, I suggest you use **shoulda-matchers** gem, please.

3- It's required that, as you can read it [here](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/friends_timeline.png), Posts timeline includes posts written by logged-in user and their friends
But take a look at this screenshot
Screenshot at 2020-11-03 16:11:29
**Bob Bar** see the **Alice Bar**'s post while they are not friends as confirmed the `Friendship` table below
```rb
2.7.1 :002 > Friendship.all
  Friendship Load (0.8ms)  SELECT  "friendships".* FROM "friendships" LIMIT $1  [["LIMIT", 11]]
 => #<ActiveRecord::Relation []> 
2.7.1 :003 > User.all
  User Load (0.6ms)  SELECT  "users".* FROM "users" LIMIT $1  [["LIMIT", 11]]
 => #<ActiveRecord::Relation [#<User id: 1, email: "alice@bar.com", created_at: "2020-11-03 15:10:53", updated_at: "2020-11-03 15:10:53", name: "Alice Bar", gravatar_url: nil>, #<User id: 2, email: "bob@bar.com", created_at: "2020-11-03 15:11:07", updated_at: "2020-11-03 15:11:07", name: "Bob Bar", gravatar_url: nil>]> 
2.7.1 :004 >
```
Kindy fix that please

4- **Bob Bar** sends a friendship request to **Alice Bar** and by clicking on the **Friends** link I got this error
Screenshot at 2020-11-03 16:14:38
Kindly, fix it, please

5- At this milestone, it's expected that you implement a **one-row friendship** feature instead of a **two-row friendship** which is required at the next milestone.
```rb
2.7.1 :002 > Friendship.all
  Friendship Load (0.6ms)  SELECT  "friendships".* FROM "friendships" LIMIT $1  [["LIMIT", 11]]
 => #<ActiveRecord::Relation [#<Friendship id: 1, user_id: 2, friend_id: 1, confirmed: true, created_at: "2020-11-03 15:14:20", updated_at: "2020-11-03 15:15:53">, #<Friendship id: 2, user_id: 1, friend_id: 2, confirmed: true, created_at: "2020-11-03 15:15:53", updated_at: "2020-11-03 15:15:53">]> 
2.7.1 :003 > 
```
The database content show us that you've implemented the two row friendship.
Kindly implement the requested feature, please

6- Read the comments left below

Kindly make the suggested changes and one done, submit a new code review request
Thanks

Mael FOSSO

**Comment by mahmoud717** [comment]
All issues have been fixed except for the fifth one. I don't see the logic in undoing some change just to redo it again later.
I know that this is a requirement, but the final project is going to have two-row friendships.

**Review by Rhelli — APPROVED** [review]
# Approved! ✅🥳

Good work on this Milestone @mahmoud717 , you've correctly implemented all the requirements and modified the project as required by the previous TSE, so good job!

As a brief note, the previous TSE was indeed correct to point out that you have added in a feature not required by this Milestone (Two-Row / Bi-Directional Friendships), however it does seem a little wasteful to scrap it only to rebuild it, so I'll let it slide. Please do pay attention the Milestone requirements in the future though as other TSEs may enforce this causing you to have to re-write code and waste time!

Nevertheless, good job on the project! Keep it up and good luck!

All the best,

💂🏽‍♂️ Rory

**Pull request merged**